file_block_encrypt_example/block_encrypt.c

```diff
@@ -0,0 +1,198 @@
+                    plaintext_buf[plaintext_buf_length];  // Variable Length
+                                                          // Arrays work from
+                                                          // C99
+
+                // Insert random bytes inside buffer.
+                RAND_bytes(plaintext_buf, plaintext_buf_length);
+
+                // If the block size exceeds MAX_PLAINTEXT_BLOCK_SIZE after
+                // adding the next block, it should be saved to the next block.
+                // The leftover space should be filled with zero padding.
+                if (total_block_size + plaintext_buf_length >
+                    MAX_PLAINTEXT_BLOCK_SIZE) {
+                    // Add zero padding to the end of the plaintext_block_buf.
+                    bzero(plaintext_block_buf + total_block_size,
+                          MAX_PLAINTEXT_BLOCK_SIZE - total_block_size);
+                    printf("Add zero paddings for the leftover %d bytes.\n",
+                           MAX_PLAINTEXT_BLOCK_SIZE - total_block_size);
+                    // Now the total_block_size becomes the
+                    // MAX_PLAINTEXT_BLOCK_SIZE.
+                    total_block_size +=
+                        MAX_PLAINTEXT_BLOCK_SIZE - total_block_size;
+                    break;
+                } else {
+                    memcpy(plaintext_block_buf + total_block_size,
+                           plaintext_buf, plaintext_buf_length);
+                    total_block_size += plaintext_buf_length;
+                }
+            }
+
+            // Save the plaintext block for future comparison.
+            fwrite(plaintext_block_buf, total_block_size, 1, plaintext_fp);
+            plaintext_file_metadata[i].block_metadata[j].length =
+                total_block_size;
+
+            // Encrypt plaintext block.
+            unsigned int encrypted_length;
+            unsigned char *encrypted = symmetric_encrypt_authenticate(
+                plaintext_block_buf, total_block_size,
+                encrypted_file_metadata[i].s_key->mac_key,
+                encrypted_file_metadata[i].s_key->mac_key_size,
+                encrypted_file_metadata[i].s_key->cipher_key,
+                encrypted_file_metadata[i].s_key->cipher_key_size, IV_SIZE,
+                &encrypted_length);
+
+            // Save the encrypted block.
+            fwrite(encrypted, encrypted_length, 1, encrypted_fp);
+            encrypted_file_metadata[i].block_metadata[j].length =
+                encrypted_length;
+            printf("Wrote encrypted block %d\n", j);
+        }
+        fclose(plaintext_fp);
+        fclose(encrypted_fp);
+        printf("Finished writing encrypted blocks to encrypted%d.txt\n", i);
+    }
+
+    ///////////////////// Decrypt and compare with plaintext
+    /////////////////////////////////////
+
+    // Read files.
+    for (int i = 0; i < TOTAL_FILE_NUM; i++) {
+        char encrypted_filename[15];
```

Comment:
Please add a separate client (file reader) to demonstrate two different entities sharing session keys. Also, make sure the file reader can read the metadata from the metadata file written by the file writer.